src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -105,35 +123,68 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
     /// be repeatedly downgraded with `allow_compaction()` to permit compaction.
     pub async fn create_sources(
         &mut self,
-        mut bindings: Vec<CreateSourceCommand<T>>,
+        mut bindings: Vec<(GlobalId, (crate::sources::SourceDesc, Antichain<T>))>,
     ) -> Result<(), StorageError> {
         // Validate first, to avoid corrupting state.
         // 1. create a dropped source identifier, or
         // 2. create an existing source identifier with a new description.
         // Make sure to check for errors within `bindings` as well.
-        bindings.sort_by_key(|b| b.id);
+        bindings.sort_by_key(|(id, _)| *id);
         bindings.dedup();
         for pos in 1..bindings.len() {
-            if bindings[pos - 1].id == bindings[pos].id {
-                Err(StorageError::SourceIdReused(bindings[pos].id))?;
+            if bindings[pos - 1].0 == bindings[pos].0 {
+                return Err(StorageError::SourceIdReused(bindings[pos].0));
             }
         }
-        for binding in bindings.iter() {
-            if let Ok(collection) = self.as_ref().collection(binding.id) {
-                let (ref desc, ref since) = collection.description;
-                if (desc, since) != (&binding.desc, &binding.since) {
-                    Err(StorageError::SourceIdReused(binding.id))?
+        for (id, description_since) in bindings.iter() {
+            if let Ok(collection) = self.as_ref().collection(*id) {
+                if &collection.description != description_since {
+                    return Err(StorageError::SourceIdReused(*id));
                 }
             }
         }
+
+        let mut dataflow_commands = vec![];
+
         // Install collection state for each bound source.
-        for binding in bindings.iter() {
-            let collection = CollectionState::new(binding.desc.clone(), binding.since.clone());
-            self.storage.collections.insert(binding.id, collection);
+        for (id, (desc, since)) in bindings {
```

Comment:
With the vector of bindings now being returned, I think it is more clear that we aren't looking at `since` when we do the work here. Like, we push a bunch of binding information that is potentially irrelevant from `since`s point of view. I'm not sure it is harmful as I'm not certain what the vector communicates, but wanted to point it out! (I think previously we .. validated nothing).